### ISSUE TITLE
Not recognizing step definition in Scenario Outline

### ISSUE THREAD

**Issue by tottinge** [rlogiacco/Natural#40]
While natural seems to do a great job with the step definition in the flows (seen above in pic), in the scenario outline, it doesn't seem to find the regex. Is it confused by the placeholders not being digits? 

![screenshot 2014-12-05 10 46 06](https://cloud.githubusercontent.com/assets/611142/5319280/6aceb74e-7c6d-11e4-92e6-6b51dccb104e.png)
##

**Comment by rlogiacco** [comment]
Duplicate of #31

**Comment by drkstr101** [comment]
Flagging for review in [next](https://github.com/rlogiacco/Natural/pull/73).

**Comment by drkstr101** [comment]
Hmm, this could be tricky. We may have to pull values from the parsed Table data to know how to match the tokens in the step definition.

**Comment by rlogiacco** [comment]
I believe we have also a visualization problem: what should we do when the first three rows match but the fourth doesn't? How to pinpoint where the problem is? Not to mention the performance degradation...

We can aim at a compromise: replace any reference to the placeholders in the step with a fixed value, let's say an **underscore**, then the step defs must be altered to include this marker char, like ```"I push [0-9_]+" ```

**Comment by drkstr101** [comment]
What if we maintain a cache of `TableRow` -> `description` transformations, where `description` contains the tokenized string replacements. Then `CucumberValidator` could inject a service in the same fassion os `IStepMatcher`,  but for the purposes of validating intances of `Example`. In this case would we not issue a warning if there are multiple valid matches, or is this an incorrct assumption about the cardinality of table data to step definitions?

Of course there would need to be a l=bit of work done at the grammer level, and, for example, providing an `IValueConverterService` implementation which canparse the table data from raw strings. I think the path is at least reachable though, no?

**Comment by rlogiacco** [comment]
When such entries are evicted from this cache and how do you identify which entries should be removed? You need to also store the source file and evict all entries in the file whenever that file is saved and rebuild that part of the cache... Then your stepmatcher must inspect the cache using the current file as an index.... 

The step matching is not blazing fast and it is doing at most one match per step, with this approach a table with 100 examples will produce match attempts x 100

**Comment by drkstr101** [comment]
Ah yes good point, I can see how this could get unwieldy.

**Comment by drkstr101** [comment]
What if we simply parse the first row (after the header), then make it clear in the warning message what description was used to validate the step definition. Then we can move the step validation check up to `AbstractScenario`, while branching on `ScenarioOutline` as a special case.

IMHO this would be more user friendly  than requiring changes to the step definitions to compensate for the editor.

**Comment by rlogiacco** [comment]
Yes, I agree with your suggestion: we only try to match the first row of the first table (multiple examples at play!) and CTRL+click brings you to the stepdef we used for matching.

This is a much better compromise than what I was suggesting. 👍
